functors_task_1/student/src/impl.cpp

```diff
@@ -2,28 +2,44 @@
 
 BeerBrand BeerOrganizer::operator()() {
     if (static_cast<size_t>(BeerBrand::Max) == currentIndex)
-        currentIndex = 1;
+        currentIndex = static_cast<size_t>(BeerBrand::HoeGaarden);
 
     return static_cast<BeerBrand>(currentIndex++);
 }
 
-bool isExpensiveBeer(const BeerBrand& brand)
+bool isExpensiveBeer(BeerBrand brand)
 {
-    std::vector<BeerBrand> expensive{BeerBrand::HoeGaarden, BeerBrand::Corona};
-
-    return std::find(expensive.begin(), expensive.end(), brand) != std::end(expensive);
+    switch(brand)
+    {
+    case BeerBrand::HoeGaarden:
+    case BeerBrand::Corona:
+        return true;
+    default:
+        return false;
+    }
 }
 
 Cocktail MixingPolicy::mix(AlcoholDrink alcohol, NonAlcoholDrink nonAlcohol)
 {
-    using Recipe = std::tuple<AlcoholDrink, NonAlcoholDrink>;
-    using Ingredients = std::pair<AlcoholDrink, NonAlcoholDrink>;
+    using Recipe = std::pair<AlcoholDrink, NonAlcoholDrink>;
+
+    struct hashPair
+    {
+        size_t operator()(const Recipe& recipe) const
+        {
+            size_t first = static_cast<size_t>(recipe.first);
+            size_t second = static_cast<size_t>(recipe.second);
+
+            size_t cantorPairing = (first + second) * (first + second + 1) / 2 + second;
```

Comment:
prefer std::hash

Reply:
Xor is giving the same result in cases of (a xor b) and (b xor a).
So for example pairs (OrangeJuice = 1, Wine = 2) and (GrapefruitJuice = 2, Whiskey = 1) will result in the same hash value which is wrong. Cantor pairing approach on the other hand handles such cases correctly.

Reply:
no problem, but make it more explicit - extract to some function, I've suggested xor for simplicity